feat: CDC ETL source

yuhao-su opened the pull request:
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://raw.githubusercontent.com/risingwavelabs/risingwave/17af8a747593ebdbfa826691daf75bdab7d14fa0/.github/contributor-license-agreement.txt).

This PR introduces the first part of the CDC ETL source, which allows you to create a etl source instead of table on shared cdc source to avoid materialized a copy of CDC data in RW tabl.

Users can use the source like any other source by creating MV, sink etc. on it.

We can run the following:
```
create source s1 with (
  ${RISEDEV_MYSQL_WITH_OPTIONS_COMMON},
  username = 'shared-cdc',
  password = 'abcdef',
  database.name = 'testdb1',
);

create source elt1_s1 (
  *
) from s1 table 'testdb1.tt1';
```

 


There are certain limitation in this PR.
TODO:
1. generated column on source definition.
2. handle parsing update messages
3. when doing backfill, need to lock the upstream database table with a read lock
4. add test for PG source

## What's changed and what's your intention?

<!--

**Please do not leave this empty!**

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:

- Summarize your change (**mandatory**)
- How does this PR work? Need a brief introduction for the changed logic (optional)
- Describe clearly one logical change and avoid lazy messages (optional)
- Describe any limitations of the current code (optional)
- Refer to a related PR or issue link (optional)

-->

## Checklist

- [ ] I have written necessary rustdoc comments.
- [ ] <!-- OPTIONAL --> I have added necessary unit tests and integration tests.
- [ ] <!-- OPTIONAL --> I have added test labels as necessary. <!-- See https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide) -->
- [ ] <!-- OPTIONAL --> I have added fuzzing tests or opened an issue to track them. <!-- Recommended for new SQL features, see #7934 -->
- [ ] <!-- OPTIONAL --> My PR contains breaking changes. <!-- If it deprecates some features, please create a tracking issue to remove them in the future -->
- [ ] <!-- OPTIONAL --> My PR changes performance-critical code, so I will run (micro) benchmarks and present the results. <!-- To manually trigger a benchmark, please check out [Notion](https://www.notion.so/risingwave-labs/Manually-trigger-nexmark-performance-dashboard-test-b784f1eae1cf48889b2645d020b6b7d3). -->
- [ ] <!-- OPTIONAL --> I have checked the [Release Timeline](https://github.com/risingwavelabs/rw-commits-history/blob/main/release_timeline.md) and [Currently Supported Versions](https://docs.risingwave.com/changelog/release-support-policy#support-end-dates-for-recent-releases) to determine which release branches I need to cherry-pick this PR into. <!-- Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md) -->


## Documentation

- [ ] <!-- OPTIONAL --> My PR needs documentation updates. <!-- Please use the **Release note** section below to summarize the impact on users -->

<details>
<summary><b>Release note</b></summary>

<!--
If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes.

Please prioritize highlighting the impact these changes will have on users.
Discuss technical details in the "What's changed" section, and focus on the impact on users in the release note.

You should also mention the environment or conditions where the impact may occur.
-->

</details>

xxchan reviewed (commented):
Can you describe what's changed in the description? Thanks

yuhao-su commented:
> Is this PR blocked by anything? Seems to be marked as WIP for a month now.

I'm still working on the following PR. I'll make it ready for review after the feature is ready for e2e

yuhao-su committed:
feat: implement CDC table source functionality

- Add parser support for CDC source creation
- Implement CDC source catalog and binder logic
- Add CDC scan optimizer and planner integration
- Include e2e tests for source from source functionality
- Add necessary migrations and proto updates

yuhao-su committed:
fix

yuhao-su committed:
fix

yuhao-su committed:
fix test

yuhao-su committed:
resolve

hzxa21 commented:
Is this PR ready for review or should we review https://github.com/risingwavelabs/risingwave/pull/22470 first? (still draft?)

chenzl25 commented:
No updates
